File: bonus/search_suggestions_system.java
class TrieNode
{
    // Can store data at nodes if desired
    Map<Character, TrieNode> children;
    List<String> suggestions;

    TrieNode()
    {
        this.children = new HashMap<>();
        this.suggestions = new ArrayList<>();
    }
}

public List<List<String>> suggestedProducts(
    String[] products, String searchWord)
{
    TrieNode root = new TrieNode();

    for (String product : products)
    {
        TrieNode node = root;
        for (char letter : product.toCharArray())
        {
            if (!node.children.containsKey(letter))
            {
                node.children.put(letter, new TrieNode());
            }
            node = node.children.get(letter);

            node.suggestions.add(product);
            Collections.sort(node.suggestions);
            if (node.suggestions.size() > 3)
            {
                node.suggestions.remove(node.suggestions.size() - 1);
            }
        }
    }

    List<List<String>> suggested_products = new ArrayList<>();
    TrieNode node = root;

    for (char letter : searchWord.toCharArray())
    {
        if (node.children.containsKey(letter))
        {
            node = node.children.get(letter);
            suggested_products.add(node.suggestions);
        }
        else
        {
            // Deadend reached
            node.children = new HashMap<>();
            suggested_products.add(new ArrayList<String>());
        }
    }

    return suggested_products;
}

File: bonus/search_suggestions_system.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.suggestions = []


def suggestedProducts(products: list[str], searchWord: str) -> list[list[str]]:
    root = TrieNode()

    for product in products:
        node = root
        for letter in product:
            if letter not in node.children:
                node.children[letter] = TrieNode()
            node = node.children[letter]

            node.suggestions.append(product)
            node.suggestions.sort()
            if len(node.suggestions) > 3:
                node.suggestions.pop()

    suggested_products = []
    node = root
    for letter in searchWord:
        if letter in node.children:
            node = node.children[letter]
            suggested_products.append(node.suggestions)
        else:
            # Deadend reached
            node.children = {}
            suggested_products.append([])

    return suggested_products
